### PR TITLE
Derived table outer scope visibility 

### PR DESCRIPTION
**Pull request by fulghum**
Add support for outer scope visibility for derived tables, as introduced in MySQL 8.0.14.

References:
- [MySQL Blog Post: Outer references for derived tables](https://dev.mysql.com/blog-archive/supporting-all-kinds-of-outer-references-in-derived-tables-lateral-or-not/)
- [MySQL Worklog task and notes for derived table outer scope visibility](https://dev.mysql.com/worklog/task/?id=461)

Resolves: https://github.com/dolthub/dolt/issues/4534

Dolt CI tests: https://github.com/dolthub/dolt/pull/4472

### CODE REVIEW

File: sql/analyzer/prune_columns.go
Comment (max-hoffman) [inline sql/analyzer/prune_columns.go:357]: i wonder if we still need this after you merge main, i tried cleaning up scalar and relational subquery analyzer lifecycles.

File: sql/analyzer/resolve_subqueries.go
Comment (max-hoffman) [inline sql/analyzer/resolve_subqueries.go:156]: might be worth considering some alternatives that expose information that more broadly applicable:
- relational subquery goes one step further to determine if it's correlated with an outer scope. you would only know this information after resolving columns.
- relation subquery has `outerCols` / `outerColsReferenced` / etc to indicate both that it is correlated, and which outer field parametrize this subquery

Comment (max-hoffman) [reply sql/analyzer/resolve_subqueries.go:156]: we want a stricter definition of cacheable depending on whether a relational subquery is correlated, not just that there is an outer scope. this info is reusable for ex: decorrelating subqueries, but also for queries like this:
```SQL
> explain select u from uv where u in (select * from (select x+1 from xy where x = u) s);
```
the above query currently panics, because we fail to realize that the relational subquery inside of the scalar subquery is correlated with outer scope, and try applying the `IndexedInSubqueryFilter` -- an invalid transformation if the subquery is not cacheable. It is important not to break similar but valid queries like the one below, which turbine depends on:
```SQL
> explain select u from uv where u in (select u+1 from uv);
+-----------------------------------------------+
| plan                                          |
+-----------------------------------------------+
| Project                                       |
|  ├─ columns: [uv.u]                           |
|  └─ IndexedInSubqueryFilter(uv.u IN ((Project |
|      ├─ columns: [(uv.u + 1)]                 |
|      └─ Table(uv)                             |
|          └─ columns: [u]                      |
|     )))                                       |
|      └─ IndexedTableAccess(uv)                |
|          └─ index: [uv.u]                     |
+-----------------------------------------------+
```

Comment (max-hoffman) [reply sql/analyzer/resolve_subqueries.go:156]: The stricter property would be recursive, also, root scope -> scalar SQ -> rel SQ -> scalar SQ -> rel SQ ... -> rel SQ with root scope reference

**Review by max-hoffman — COMMENTED** [review]
Maintaining cacheability when a relational subquery is not correlated might be necessary for this PR. I worry some of the integration query plans will be broken otherwise.

File: sql/analyzer/prune_columns.go
Comment (fulghum) [reply sql/analyzer/prune_columns.go:357]: I double checked this one and confirmed that this check is still needed, even with the latest changes from main.

File: sql/analyzer/resolve_subqueries.go
Comment (fulghum) [reply sql/analyzer/resolve_subqueries.go]: I added `SubqueryAlias.CanCacheResults`, to match `Subquery.CanCacheResults`. That gets us back to the same caching behavior as before, although I did find a few places where we were previously caching a derived table that shouldn't have been cached (e.g. a derived table that had a `RAND()` expression in it was previously cached, but now is not eligible for caching because of the non-determinate expression). 

I just tried out both of those queries you suggested and confirmed they both execute correctly on GMS now and match MySQL's behavior. Happy to make sure those are included in the test suite this afternoon, too.

**Comment by fulghum** [comment]
Thanks for the feedback @max-hoffman. I've got subquery alias caching working correctly now and all tests are passing. I'm still waiting on Dolt CI tests to run (https://github.com/dolthub/dolt/pull/4472), but they're looking good so far. 

I did have to add one hack that I would still like to unwind... I changed `CachedResults.WithChildren` to **not** return a new instance, in order to fix the cache mem leaks I was seeing. That obviously isn't a good change, but I figured it might be okay for a short amount of time while we fix the CachedResults mem cache leak. 

I'm going to give the code another review myself this afternoon and make sure there aren't any TODOs left in there or anything else to clean up, but other than that I think it's ready for you and @zachmu to comb over and see if you see any concerns.

File: sql/analyzer/resolve_subqueries.go
Comment (zachmu) [inline sql/analyzer/resolve_subqueries.go:157]: Probably prefer a method on Scope for this logic (ditto the other place this is done)

File: sql/analyzer/resolve_subqueries.go
Comment (zachmu) [inline sql/analyzer/resolve_subqueries.go:52]: Isn't this transform.NodeExprs? Also eliminated the sql.Expressioner check

File: sql/analyzer/resolve_subqueries.go
Comment (zachmu) [inline sql/analyzer/resolve_subqueries.go:297]: Same comment about using transform.NodeExprs (or OneNodeExprsWithNode if you don't want it recursively applied)

File: sql/analyzer/scope.go
Comment (zachmu) [inline sql/analyzer/scope.go:35]: Worried about this extra bit of state -- who's responsible for setting it back to false?

File: sql/cache.go
Comment (zachmu) [inline sql/cache.go:136]: This has a lot of performance overhead. Do we need this to distinguish null entries in the cache?

**Review by zachmu — APPROVED** [review]
Mostly looks good.

Main comment is that we probably do need to solve the cacheability problem now, as it's a major performance problem for at least one customer to lose it.

Probably makes sense to introduce some simpler queries to test cacheability that reproduce the current problem.

File: sql/analyzer/scope.go
Comment (fulghum) [reply sql/analyzer/scope.go:35]: The `Scope.newScope` method handles creating a new subscope from the current scope and that method does not set this flag on the new subscope, so it gets correctly cleared for each new subscope.

File: sql/cache.go
Comment (fulghum) [reply sql/cache.go:136]: This was needed to match the existing behavior of the `lruCache` and `historyCache` implementations. They both return `ErrKeyNotFound` when the key isn't found, and we have some other code (in Subquery.Eval if I remember correctly) that depends on that behavior to work correctly. I believe we can trigger a bug with the current code today under the right situation, since it won't detect nil as not found.

Comment (fulghum) [reply sql/cache.go:136]: If perf from looking at the error response is a concern, it of course probably wouldn't be too hard to change the API expectations to not return that error.

File: enginetest/queries/query_plans.go
Comment (fulghum) [reply enginetest/queries/query_plans.go:7301]: Yup yup, I've been talking to Max about this and we all are in agreement that we need to keep that HashLookup node. Max is in there reworking the HashLookup support, so he offered to dig into this one more and make sure it continues to get applied correctly. He's already pulled this into his dev branch and is using it, so I suspect the right way to get this merged out to main is through his join planning rewrite and we'll make sure the HashLookup is present there. I'll follow up with him on the plan for getting this to main and make sure he's on the same page.

File: sql/cache.go
Comment (fulghum) [reply sql/cache.go:136]: I dug through usage of `KeyValueCache.Get` some more this morning and there are several places where the code does need to distinguish between cached with a result of nil and not present. I think we still want this change to make the `mapCache` implementation work the same as the other implementations, but let me know if you think refactoring the interface for the cache implementations to not use an error for this signal needs to happen as part of this PR.

**Review by zachmu — APPROVED** [review]
LGTM!

**Comment by max-hoffman** [comment]
I found a bug where the local cache is filled and successfully copied to the global cache, but there is only one primary row and the next step of copying the global cache into HashLookup and deleting the global cache is never triggered. This got me thinking more about the whole cache setup we have going on. Afaik the handoff is error proof right now, but is the handoff complexity necessary? Should the *sql.Context just be responsible for cache lifecycle if there is always a global component? I think your cache identifier and global ids are more dev-friendly than opaque disposal functions (maybe a combination of session Id plus incrementing integer would be a slight improvement for multi-user sessions). Nodes communicating cache references by id, and then using `get()` and `dispose()` into the context-managed caches all the way through just seems simpler?

File: enginetest/queries/query_plans.go
Comment (fulghum) [reply enginetest/queries/query_plans.go]: Just closing the loop here... the latest commits on this PR correctly add back the `HashLookup` node for this subquery.

File: enginetest/queries/query_plans.go
Comment (fulghum) [inline enginetest/queries/query_plans.go:929]: This `HashLookup` is not getting applied because of an existing bug in RightJoin planning that I uncovered in this PR. Max also found issues with RightJoins has rewritten RightJoin planning in his PR, which will fix this one and enable proper HashLookup again here.

**Pull request merged**
